### PR TITLE
Better Block Message Support

### PR DESCRIPTION
**Pull request by tonisevener**
**Phabricator:** 
https://phabricator.wikimedia.org/T169013
https://phabricator.wikimedia.org/T275118
https://phabricator.wikimedia.org/T201640

### Notes
This PR adds improved display of blocked messages to users when creating an account, publishing an article description, and publishing an edit in our section editor. It adjusts the API calls in these areas and parses the error responses. It also adds a new blocked message panel (from our existing Panels.swift set of views) and disables the editors when the user is blocked.

This PR is best reviewed commit-by-commit. I tried to leave detail in the commit messages. Any developer testing should be tested with all commits in place. 

There are a couple of things remaining after this PR for blocked messages:

1. Showing a confirmation if the user is going to lose their edits after tapping a link in the panel, which will be integrated after #4454 is merged.
2. We have some url routing issues with some of the blocked messages, like our native talk page intercepting a link like https://en.wikipedia.org/wiki/Special:MyTalk. I put up a couple of temporary fixes for EN, but a better approach needs to be in place for other language support (see Phab [task](https://phabricator.wikimedia.org/T328678)).

### Test Steps
1. Block one of your accounts on [Test Wikipedia](https://test.wikipedia.org/wiki/Special:Block) (if you don't have admin privileges, I can block one for you) and on [Test Wikidata](https://test.wikidata.org/wiki/Special:Block).
2. Log into that account on device. Set up your primary app language as non-Test language (this is so that the correct MediaWiki:Blockedtext template is fetched after blocked API error. Test Wikipedia doesn't have one set up).
4. Visit an article on Test Wikipedia. Go to section editor. Confirm blocked panel appears after landing on editor, and editor is disabled.
5. Go to article description editor by tapping the first pencil (you may need to hunt around for a Test article with a description already in place...there's something on the PCS-side that doesn't show the "Add article description" button for those without).
6. Confirm blocked panel appears after landing on article description editor, and description editing is disabled.
7. Temporary adjust code so that article description editor updates on Wikidata rather than on Test Wikipedia. The changes in [this temp branch commit](https://github.com/wikimedia/wikipedia-ios/commit/ac6850f7d2912e8fe9d199659b6c30647427e51c) will do it.
8. Repeat steps 4 & 5. Confirm blocked panel still appears.
9. Brief regression testing - unblock account on Test Wikipedia and confirm you can successfully publish an edit in section editor.
10. Now test create account blocked error. Set up app languages to only contain Test Wikipedia. Then re-block account, making sure there's also an IP block in place.
11. Try to create an account from our create account screen. Confirm blocked message appears after attempting to create an account.

_Note: I also have blocked panels showing upon Publish, not just landing on the editors. Test steps for this will be a bit tedious - basically repeating 1-8, but unblocking your account in between each step so you can get past disabling of the editors. I didn't include these steps in the interest of leaning on QA more for this._

### Screenshots

Create Account Error
IMG_0157

Section Editor Error
IMG_0156

### CODE REVIEW

File: Wikipedia/Code/WikidataFetcher.swift [hunk 1]
```diff
@@ -51,7 +16,115 @@ public enum ArticleDescriptionSource: String {
...
 @objc public final class WikidataFetcher: Fetcher {
+    
+// MARK: Get Blocked Info Models & Methods
+    
+    public struct WikidataErrorsResult: Decodable {
+
+        struct Query: Codable {
+            struct Page: Codable {
+                let pageid: Int?
+                let ns: Int?
```
Comment (staykids) [inline Wikipedia/Code/WikidataFetcher.swift:27]: These two currently unused.

**Review by staykids — COMMENTED** [review]
Nice – definitely more involved than I initially expected! 

I haven't yet tested in use beyond confirming basic presentation of the block panel when section editing from a blocked account, but below is my first pass of notes from pure code review.

File: Wikipedia/assets/codemirror/codemirror-index.html [hunk 2]
```diff
@@ -452,6 +453,7 @@
           }, 
           inputStyle: 'contenteditable',
           direction: direction,
+		  readOnly: setupReadOnly(readOnly),
```
[the remark this answers is not in the document]
Comment (tonisevener) [reply Wikipedia/assets/codemirror/codemirror-index.html]: Done!

File: Wikipedia/Code/ScrollableEducationPanelViewController.swift [hunk 3]
```diff
@@ -203,13 +180,48 @@ class ScrollableEducationPanelViewController: UIViewController, Themeable {
...
+                                                                    ])
+        
+        var attributes: [NSAttributedString.Key : Any] = [:]
+        if let subheadingParagraphStyle = subheadingParagraphStyle {
+            attributes[NSAttributedString.Key.paragraphStyle] = subheadingParagraphStyle
+        }
+        attributedText.addAttributes(attributes, range: NSRange(location: 0, length: attributedText.length))
+        
+        subheadingTextView.attributedText = attributedText.removingRepetitiveNewlineCharacters()
+        subheadingTextView.tintColor = theme.colors.link
```
[the remark this answers is not in the document]
Comment (tonisevener) [reply Wikipedia/Code/ScrollableEducationPanelViewController.swift]: Good catch, done!

File: Wikipedia/Code/SectionFetcher.swift [hunk 4]
```diff
@@ -1,42 +1,54 @@
...
-                case contentmodel
-                case contentformat
-                case asterisk = "*"
+            struct Page: Codable {
+                struct Revision: Codable {
+                    struct Slot: Codable {
+                        let contentmodel: String?
+                        let contentformat: String?
+                        let content: String?
+                        enum CodingKeys: String, CodingKey {
```
[the remark this answers is not in the document]
Comment (tonisevener) [reply Wikipedia/Code/SectionFetcher.swift:16]: Good point, I decided to go the opposite direction and define more CodingKeys to make more of these properties camel case.

File: Wikipedia/Code/SectionFetcher.swift [hunk 5]
```diff
@@ -1,42 +1,54 @@
...
+                }
+                
+                struct Protection: Codable {
+                    let type: String?
+                    let level: String?
+                    let expiry: String?
+                }
+                
+                let pageid: Int?
+                let ns: Int?
```
[the remark this answers is not in the document]
Comment (tonisevener) [reply Wikipedia/Code/SectionFetcher.swift]: Removed!

File: Wikipedia/Code/WikidataFetcher.swift (hunk 1 above)
Comment (tonisevener) [reply Wikipedia/Code/WikidataFetcher.swift]: Removed!

**Review by staykids — APPROVED** [review]
Nice! Was able to test most edit/creation block paths after I got Admin rights on test Wiki (other than Wikidata blocks), but as mentioned perhaps best to rely on QA for exhaustive testing.

Looks like just the conflict with our shared changes (sorry!) to the section editor to resolve.

**Comment by tonisevener** [comment]
@staykids thanks! Merged conflict is fixed. To merge our two features, I changed it so that if a blocked message does return from the wikitext call, we disable the automatic edit notice presentation (I feel there's little need to show it to them if the editor is disabled). Writing up a Phab comment now for design to confirm this thinking.

It was difficult to test because I doubt very many Test Wikipedia articles have edit notices. I basically did it by mapping the edit notice response locally with a response taken from EN Wikipedia.

**Comment by staykids** [comment]
> @staykids thanks! Merged conflict is fixed. To merge our two features, I changed it so that if a blocked message does return from the wikitext call, we disable the automatic edit notice presentation (I feel there's little need to show it to them if the editor is disabled). Writing up a Phab comment now for design to confirm this thinking.
> 
> It was difficult to test because I doubt very many Test Wikipedia articles have edit notices. I basically did it by mapping the edit notice response locally with a response taken from EN Wikipedia.

@tonisevener  Cool - I tested by just stubbing out an always successful/available but empty edit notice in the section editor, and the behavior appears as you mention where the block alert supersedes the modal presentation, but the modal's still available via toolbar button.

I'll go ahead and merge so this doesn't fall too far off `main`, and we can adjust based on design feedback if needed.

**Pull request merged**